theseus/optimizer/nonlinear/dogleg.py

```diff
@@ -0,0 +1,106 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+from typing import Any, Dict, Optional, Type
+
+import torch
+
+from theseus.core import Objective
+from theseus.optimizer import Linearization
+from theseus.optimizer.linear import LinearSolver
+
+from .trust_region import TrustRegion
+
+
+# See Nocedal and Wright, Numerical Optimization, pp. 73-77
+# https://www.csie.ntu.edu.tw/~r97002/temp/num_optimization.pdf
+class Dogleg(TrustRegion):
+    EPS = 1e-7
+
+    def __init__(
+        self,
+        objective: Objective,
+        linear_solver_cls: Optional[Type[LinearSolver]] = None,
+        vectorize: bool = False,
+        linearization_cls: Optional[Type[Linearization]] = None,
+        linearization_kwargs: Optional[Dict[str, Any]] = None,
+        linear_solver_kwargs: Optional[Dict[str, Any]] = None,
+        abs_err_tolerance: float = 1e-10,
+        rel_err_tolerance: float = 1e-8,
+        max_iterations: int = 20,
+        step_size: float = 1.0,
+        **kwargs,
+    ):
+        super().__init__(
+            objective,
+            linear_solver_cls=linear_solver_cls,
+            vectorize=vectorize,
+            linearization_cls=linearization_cls,
+            linearization_kwargs=linearization_kwargs,
+            linear_solver_kwargs=linear_solver_kwargs,
+            abs_err_tolerance=abs_err_tolerance,
+            rel_err_tolerance=rel_err_tolerance,
+            max_iterations=max_iterations,
+            step_size=step_size,
+            **kwargs,
+        )
+
+    def _compute_delta_impl(self) -> torch.Tensor:
+        trust_region_2 = self._trust_region**2
+        delta_gn = self.linear_solver.solve()
+        delta_gn_norm_2 = TrustRegion._squared_norm(delta_gn)
+        # All Gauss-Newton step are within trust-region, can return
+        if (delta_gn_norm_2 < trust_region_2).all():
+            # Return a False mask since no indices are at the boundary
+            return delta_gn
+
+        # ---------------------------------------------------------------------
+        # Some Gauss-Newton steps are outside trust-region,
+        # need to compute dogleg step
+        # ---------------------------------------------------------------------
+
+        linearization = self.linear_solver.linearization
+
+        delta_sd = linearization.Atb.squeeze(2)
+        Adelta_sd = linearization.Av(delta_sd)
+        Adelta_sd_norm_2 = TrustRegion._squared_norm(Adelta_sd)
+        grad_norm_2 = TrustRegion._squared_norm(delta_sd)
+        cauchy_step_size = grad_norm_2 / (Adelta_sd_norm_2 + Dogleg.EPS)
+        delta_c = delta_sd * cauchy_step_size
+        delta_c_norm_2 = TrustRegion._squared_norm(delta_c)
```

Comment:
Maybe simplify the computation as `delta_c_norm_2 = grad_norm_2*Cauchy_step_size**2`?

Reply:
Good catch!